### DISCUSSION TITLE
[Snowflake] Object 'CHANGES' does not exist or not authorized. (SQL-42S02)

### DISCUSSION THREAD

**Discussion by qqcvd** [theory/sqitch#694] [Q&A] [answered]
Hello,
I am new on sqitch. The doc says "First Sqitch created registry tables used to track database changes." but I have the followinf error :
Object 'CHANGES' does not exist or not authorized. (SQL-42S02)

I looked into the QUERY_HISTORY table and I see the SELECT error but There is no creation request before.
Do you know what is the correct process to init the CHANGES table?

Regards
Stephane

**Comment by theory** [comment]
Creation comes after the query failure; make sure the user you're connecting as has permission to create the schema and tables.

**Reply by qqcvd** [reply]
Hello David
I understand that Creation comes after the query failure, but i see that in the Snowflake QUERY_HISTORY, there is no creation request after the Select failure.

**Comment by theory** [comment]
Does the user you're connecting as have permission to create the necessary schema and tables? If you don't see a `CREATE SCHEMA` statement in the logs, perhaps it doesn't log it if the user doesn't have permission?

**Reply by qqcvd** [reply]
Hello,
The user is the owner of the database, 
I understood the probleme this morning : Sqith both uses ODBC connection and Snowql connection.
ODBC connection settings are taken from sqitch.conf file but Snowql connection settings are taken from ~/.snowsql/config file.
This is kind of a problem because ~/.snowsql/config is kind of general default config file while we need specific config setting for current sqitch project.
209 / 5 000
Résultats de traduction
Résultats de traduction
Do you know how to set Snowsql parameters on Sqitch? I don't know where I can specify these parameters, especially since we don't have control over the Snowsql command line executed by Sqitch.

Regards 
Stéphane

**Comment by theory** [comment]
Sqitch reads select values from `~/.snowsql.config`. Details [here](https://sqitch.org/docs/manual/sqitch-authentication/). The relevant bit:

> The variables that Sqitch currently reads are:
> 
> *   connections.accountname
> *   connections.username
> *   connections.password
> *   connections.rolename
> *   connections.region (Deprecated by Snowflake)
> *   connections.warehousename

**Reply by qqcvd** [reply]
Hi
Thanks for your help and the link and the explanations. 
To my mind, this is not ideal because it requires changing the contents of the general Snowsql settings file when deploying to the next environment, but I found another solution : we can set unix environment variables which i store in a file to execute before executing the Sqitch Deploy command.
This is what my  myDevEnvironment.sh file might look like :

``` sh
export SNOWSQL_ACCOUNT=MySnowflakeAccountName
export SNOWSQL_USER=USER_SQITCH
export SNOWSQL_PWD="SqitchPassword"
export SNOWSQL_DATABASE=DB_DEV_HUB
export SNOWSQL_ROLE=RL_DEV_SQITCH
export SNOWSQL_WAREHOUSE=WH_DEV_SQITCH
```

The point is that it seems redundant with the sqitch/conf file 
Do you know why Sqitch uses 2 connection types (ODBC and Snowql) ?
Regards 
Stéphane

**Comment by theory** [comment, marked as answer]
I _do_ know why! It's because it uses a native connection in the code to manage the registry tables (changes, log, tags, etc.), but delegates running your change scrips to the command-line client native to your platform (SnowSQL in your case).

Using environment variables sounds like the right way to go for you. Consider using a tool like [direnv](https://direnv.net) in your environment to manage project-specific variables, and be I recommend that you be very specific about defining variables in you CI/CD pipelines. You can find a complete list of supported environment variables in [sqitch help environment](https://sqitch.org/docs/manual/sqitch-environment/).
